src/deploy/functions/containerCleaner.ts

```diff
@@ -128,6 +129,112 @@ export class ContainerRegistryCleaner {
+ * @param dockerHelpers: a map of {@link SUBDOMAINS} to {@link DockerHelper}. If omitted, will use the default value and create each {@link DockerHelper} internally.
+ *
+ * @throws {@link FirebaseError}
+ * Thrown if the provided location is not a valid Google Cloud region or we fail to search subdomains.
+ */
+export async function listGcfPaths(
+  projectId: string,
+  locations?: string[],
+  dockerHelpers: Record<string, DockerHelper> = {}
+): Promise<string[]> {
+  if (!locations) {
+    locations = Object.keys(SUBDOMAIN_MAPPING);
+  }
+  const invalidRegion = locations.find((loc) => !SUBDOMAIN_MAPPING[loc]);
+  if (invalidRegion) {
+    throw new FirebaseError(`Invalid region ${invalidRegion} supplied`);
+  }
+  const locationsSet = new Set(locations); // for quick lookup
+  const subdomains = new Set(Object.values(SUBDOMAIN_MAPPING));
+  const failedSubdomains = new Set<string>();
+
+  const listAll = [...subdomains].map((subdomain) => {
+    try {
+      return getHelper(dockerHelpers, subdomain).ls(`${projectId}/gcf`);
+    } catch (err) {
+      failedSubdomains.add(subdomain);
+      logger.debug(err);
+      const stat: Stat = {
+        children: [],
+        digests: [],
+        tags: [],
+      };
+      return Promise.resolve(stat);
+    }
+  });
+
+  const gcfDirs = (await Promise.all(listAll))
+    .map((results) => results.children)
+    .reduce((acc, val) => [...acc, ...val], [])
+    .filter((loc) => locationsSet.has(loc));
+
+  if (failedSubdomains.size == subdomains.size) {
+    throw new FirebaseError("Failed to search all subdomains.");
+  } else if (failedSubdomains.size > 0) {
+    const failed: string = [...failedSubdomains].join(",");
+    throw new FirebaseError(`Failed to search the following subdomains: ${failed}`);
+  }
+
+  return gcfDirs.map((loc) => {
+    return `${SUBDOMAIN_MAPPING[loc]}.${containerRegistryDomain}/${projectId}/gcf/${loc}`;
+  });
+}
+
+/**
+ * Deletes all artifacts from GCF directory in GCR.
+ * @param projectId: the current project that contains GCF artifacts
+ * @param location: the specific region to be clean up. If omitted, will delete all locations.
+ * @param dockerHelpers: a map of {@link SUBDOMAINS} to {@link DockerHelper}. If omitted, will use the default value and create each {@link DockerHelper} internally.
+ *
+ * @throws {@link FirebaseError}
+ * Thrown if the provided location is not a valid Google Cloud region or we fail to delete subdomains.
+ */
+export async function deleteGcfArtifacts(
+  projectId: string,
+  locations?: string[],
+  dockerHelpers: Record<string, DockerHelper> = {}
+): Promise<void> {
+  if (!locations) {
+    locations = Object.keys(SUBDOMAIN_MAPPING);
+  }
+  const invalidRegion = locations.find((loc) => !SUBDOMAIN_MAPPING[loc]);
+  if (invalidRegion) {
+    throw new FirebaseError(`Invalid region ${invalidRegion} supplied`);
+  }
+  const subdomains = new Set(Object.values(SUBDOMAIN_MAPPING));
+  const failedSubdomains = new Set<string>();
```

Comment:
arrays are idiomatic in JavaScript and you're only appending to the end. You can use an array and just use push(). Array insertion into an end is O(1) when using the doubling method.

Reply:
Changed to an array.